File: mobile/lib/utils/error_handler.dart
/// Fonction utilitaire pour extraire un message d'erreur formaté depuis une exception
/// 
/// Cette fonction extrait le message formaté avec solutions depuis les exceptions
/// lancées par ApiService._handleError().
/// 
/// @param error - L'erreur capturée (peut être de type dynamic)
/// @param fallbackMessage - Message par défaut si aucun message n'est trouvé
/// @returns Le message d'erreur formaté avec solutions
/// 
/// @example
/// ```dart
/// try {
///   await apiService.getForms();
/// } catch (e) {
///   final message = getErrorMessage(e, 'Une erreur est survenue');
///   ScaffoldMessenger.of(context).showSnackBar(
///     SnackBar(content: Text(message)),
///   );
/// }
/// ```
String getErrorMessage(dynamic error, [String fallbackMessage = 'Une erreur est survenue']) {
  if (error == null) {
    return fallbackMessage;
  }
  
  // Si c'est une Exception, extraire le message directement
  if (error is Exception) {
    final message = error.toString().replaceFirst('Exception: ', '');
    return message.isNotEmpty ? message : fallbackMessage;
  }
  
  // Si c'est une String, l'utiliser directement
  if (error is String) {
    return error.isNotEmpty ? error : fallbackMessage;
  }
  
  // Sinon, convertir en String
  final message = error.toString();
  return message.isNotEmpty ? message : fallbackMessage;
}

/// Vérifie si une erreur est une erreur réseau (pas de réponse du serveur)
bool isNetworkError(dynamic error) {
  final message = getErrorMessage(error, '').toLowerCase();
  return message.contains('timeout') ||
         message.contains('connexion') ||
         message.contains('réseau') ||
         message.contains('connection') ||
         message.contains('network');
}

/// Vérifie si une erreur nécessite une reconnexion (401, 403)
bool requiresReauth(dynamic error) {
  final message = getErrorMessage(error, '').toLowerCase();
  return message.contains('session expirée') ||
         message.contains('401') ||
         message.contains('403') ||
         message.contains('reconnecter');
}

File: frontend/src/utils/error-handler.ts
/**
 * Fonction utilitaire pour extraire un message d'erreur formaté depuis une erreur API
 * 
 * Cette fonction utilise le champ `userMessage` ajouté par l'intercepteur axios
 * qui contient des messages détaillés avec des solutions suggérées.
 * 
 * @param error - L'erreur capturée (peut être de type any)
 * @param fallbackMessage - Message par défaut si aucun message n'est trouvé
 * @returns Le message d'erreur formaté avec solutions
 * 
 * @example
 * ```ts
 * try {
 *   await api.get('/endpoint');
 * } catch (error) {
 *   const message = getErrorMessage(error, 'Une erreur est survenue');
 *   showAlert('Erreur', message);
 * }
 * ```
 */
export function getErrorMessage(error: any, fallbackMessage: string = 'Une erreur est survenue'): string {
  // Priorité 1: Message formaté avec solutions (ajouté par l'intercepteur axios)
  if (error?.userMessage) {
    return error.userMessage;
  }
  
  // Priorité 2: Message du serveur dans la réponse
  if (error?.response?.data?.message) {
    return error.response.data.message;
  }
  
  // Priorité 3: Message d'erreur dans response.data.error
  if (error?.response?.data?.error) {
    return error.response.data.error;
  }
  
  // Priorité 4: Message d'erreur standard
  if (error?.message) {
    return error.message;
  }
  
  // Fallback: Message par défaut
  return fallbackMessage;
}

/**
 * Vérifie si une erreur est une erreur réseau (pas de réponse du serveur)
 */
export function isNetworkError(error: any): boolean {
  return !error?.response && !!error?.request;
}

/**
 * Vérifie si une erreur est une erreur de timeout
 */
export function isTimeoutError(error: any): boolean {
  return error?.code === 'ECONNABORTED' || 
         error?.message?.includes('timeout') ||
         error?.response?.status === 504;
}

/**
 * Vérifie si une erreur nécessite une reconnexion (401, 403)
 */
export function requiresReauth(error: any): boolean {
  return error?.response?.status === 401 || error?.response?.status === 403;
}

